### PR TITLE
[MEN-5351] CLI to sync devices to Cloud providers

### PR DESCRIPTION
**Pull request by alfrunes**

### CODE REVIEW

**Comment by alfrunes** [comment]
:repeat_one: @mender-test-bot start pipeline

File: store/datastore.go [hunk 1]
```diff
@@ -42,6 +42,9 @@ type DataStore interface {
 	DoDevicesExistByIntegrationID(context.Context, uuid.UUID) (bool, error)
 	// RemoveDevicesFromIntegration integration with given id from all devices
 	// belonging to the tenant within the context.
+	RemoveIntegrationFromDevices(ctx context.Context, id uuid.UUID) (count int64, err error)
+	// UsertDeviceIntegrations adds the list of integration IDs to the
+	// device and creates it if it does not exist.
```
Comment (merlin-northern) [inline store/datastore.go:47]: RemoveDevicesFromIntegration does not match `RemoveIntegrationFromDevices` and a small misspell in UsertDeviceIntegrations

File: store/datastore.go [hunk 2]
```diff
@@ -50,6 +53,15 @@ type DataStore interface {
 	DeleteDevice(ctx context.Context, deviceID string) error
 	SetIntegrationCredentials(context.Context, uuid.UUID, model.Credentials) error
 	RemoveIntegration(context.Context, uuid.UUID) error
+
+	// GetAllDevices returns an iterator over ALL devices sorted by tenant ID.
```
Comment (merlin-northern) [inline store/datastore.go:57]: I do not entirely follow sorting by tenant ID, it is just a dictionary asc sort?

File: store/mongo/datastore_mongo.go [hunk 3]
```diff
@@ -496,9 +522,19 @@ func (db *DataStoreMongo) UpsertDeviceIntegrations(
...
+
+func (db *DataStoreMongo) GetAllDevices(ctx context.Context) (store.Iterator, error) {
+	collDevs := db.Collection(CollNameDevices)
+
+	return collDevs.Find(ctx,
+		bson.D{},
+		mopts.Find().
+			SetSort(bson.D{{Key: KeyTenantID, Value: 1}}),
+	)
+
```
Comment (merlin-northern) [inline store/mongo/datastore_mongo.go:539]: nit: do we need this extra empty line here?

File: store/datastore.go (hunk 1 above)
Comment (alfrunes) [reply store/datastore.go:47]: LOL, the description is right though :smile:

File: store/datastore.go (hunk 2 above)
Comment (alfrunes) [reply store/datastore.go:57]: I've utilized the fact that the devices are returned sorted (order doesn't really matter) in the caching for the CLI command.

File: store/mongo/datastore_mongo.go (hunk 3 above)
Comment (alfrunes) [reply store/mongo/datastore_mongo.go:539]: Nope, that's an unintended fmt :)

File: app/app.go [hunk 4]
```diff
@@ -253,6 +255,10 @@ func (a *app) DeleteIOTHubDevice(ctx context.Context, deviceID string) error {
 			}
 			err = a.hub.DeleteDevice(ctx, cs, deviceID)
 			if err != nil {
+				if htErr, ok := err.(client.HTTPError); ok &&
+					htErr.Code() == http.StatusNotFound {
+					continue
```
Comment (tranchitella) [inline app/app.go:411]: Not super important, but this is not covered by unit tests.

File: app/app.go [hunk 5]
```diff
@@ -240,6 +244,139 @@ func (a *app) ProvisionDevice(
...
+		}
+	}
+
+	return nil
+}
+
+func (a app) syncCacheIntegrations(ctx context.Context) (map[uuid.UUID]*model.Integration, error) {
+	// Cache integrations for the given tenant
+	integrations, err := a.store.GetIntegrations(
+		ctx, model.IntegrationFilter{Limit: 20},
```
Comment (tranchitella) [inline app/app.go:318]: Why 20? By the way, maybe it should be a constant.

File: main.go [hunk 6]
```diff
@@ -70,13 +82,35 @@ func doMain(args []string) {
...
 			},
+			{
+				Name:   "sync-devices",
+				Usage:  "Synchronize device state across IoT platforms.",
+				Action: cmdSync,
+				Flags: []cli.Flag{
+					&cli.IntFlag{
+						Name:  "batch-size",
+						Usage: "Maximum number of devices to sync in a batch.",
+						Value: 50,
```
Comment (tranchitella) [inline main.go:93]: Can we set a min and max value for this parameter? What is going to happen if I set an arbitrarily large number?
Or a negative number?

**Review by tranchitella — APPROVED** [review]
The PR is massive, looks good to me, I have a few questions but nothing really important.

File: client/iothub/client.go [hunk 7]
```diff
@@ -250,30 +250,36 @@ func (c *client) DeleteDevice(ctx context.Context, cs *model.ConnectionString, i
 }
 
 func (c *client) GetDeviceTwins(
-	ctx context.Context, cs *model.ConnectionString,
-) (Cursor, error) {
-	const (
-		SQLQuery              = `{"query":"SELECT * FROM devices"}`
-		pageSize, pageSizeStr = 100, "100"
+	ctx context.Context, cs *model.ConnectionString, deviceIDs []string,
```
Comment (kjaskiewiczz) [inline client/iothub/client.go:253]: is there a limit for number of devices iothub can return?

File: app/app.go [hunk 8]
```diff
@@ -240,6 +244,139 @@ func (a *app) ProvisionDevice(
...
+
+		switch integration.Provider {
+		case model.ProviderIoTHub:
+			err := a.syncIoTHubDevices(ctx, deviceIDs, *integration)
+			if err != nil {
+				return err
+			}
+		default:
+			// Invalid integration
+			// FIXME(alf) what to do?
```
Comment (kjaskiewiczz) [inline app/app.go:308]: planning to fix?

File: client/devauth/client.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+		http.MethodGet,
+		c.uri+repl.Replace(URIInternalDevices),
+		nil,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "devauth: error preparing request")
+	}
+	q := url.Values{
+		"id":       deviceIDs,
+		"per_page": []string{strconv.Itoa(len(deviceIDs))},
```
[the remark this answers is not in the document]
Comment (kjaskiewiczz) [reply client/devauth/client.go:171]: after reviewing rest of the PR, I can see is handled in the command, so I'm leaving the decision to you, if we should handle this case in the deviceauth client or not

File: app/app.go (hunk 4 above)
Comment (alfrunes) [reply app/app.go:411]: I'll follow up in my upcoming PR with acceptance tests.

File: app/app.go (hunk 5 above)
Comment (alfrunes) [reply app/app.go:318]: Right now, we don't allow more than 1 anyway, so I just picked a "reasonable" number.

File: main.go (hunk 6 above)
Comment (alfrunes) [reply main.go:93]: There is a check down in `cmdSync` forcing batch size to be between 1 and 500.

File: client/iothub/client.go (hunk 7 above)
Comment (alfrunes) [reply client/iothub/client.go:253]: I think it is `1000`, but the documentation doesn't really say, but this is the maximum number of devices returned by the `GET /devices`. I will test with 500 just to be sure.

File: app/app.go (hunk 8 above)
Comment (alfrunes) [reply app/app.go:308]: To be honest, I don't really know what to do as this case shouldn't happen. Do you think we can ignore it for now?

File: client/devauth/client.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (alfrunes) [reply client/devauth/client.go:171]: Right now I think deviceauth will return status code 400 if we exceed the limit which I think is OK.

File: client/iothub/client.go (hunk 7 above)
Comment (alfrunes) [reply client/iothub/client.go:253]: I can confirm, the API indeed allows more than 500 items:
```
$ curl https://[...].azure-devices.net/devices/query?api-version=2021-04-12 [...] -d '{"query": "SELECT * FROM devices"}' -H 'x-ms-max-item-count: 501' | jq '. | length'
501
```

File: app/app.go (hunk 8 above)
Comment (kjaskiewiczz) [reply app/app.go:308]: I think it's fine to ignore it

**Review by kjaskiewiczz — APPROVED** [review]
nice work!

**Pull request merged**